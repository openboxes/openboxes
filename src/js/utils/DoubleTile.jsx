import React from 'react';

import PropTypes from 'prop-types';
import { RiInformationLine } from 'react-icons/ri';

import useTranslate from 'hooks/useTranslate';
import CustomTooltip from 'wrappers/CustomTooltip';

const DoubleTile = ({
  cardTitle,
  cardTitleDefaultValue,
  cardFirstValue,
  cardSecondValue,
  cardFirstSubtitle,
  cardDefaultFirstSubtitle,
  cardSecondSubtitle,
  cardDefaultSecondSubtitle,
  cardInfo,
  cardInfoDefaultValue,
  currencyCode,
  formatSecondValueAsCurrency,
}) => {
  const translate = useTranslate();

  const formatCurrency = (value) => {
    if (value >= 1000000) {
      return `${(value / 1000000).toFixed(3)} ${translate('react.default.million.label', 'million')} ${currencyCode}`;
    }

    return `${new Intl.NumberFormat('en-US', {
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    }).format(value)} ${currencyCode}`;
  };

  return (
    <div className="double-tile">
      <span className="double-tile__title">
        {translate(cardTitle, cardTitleDefaultValue || cardTitle)}
      </span>
      <div className="d-flex">
        <div className="double-tile__first-value-block">
          <span className="double-tile__value-number">
            {cardFirstValue}
          </span>
          <span className="double-tile__subtitle">
            {translate(cardFirstSubtitle, cardDefaultFirstSubtitle)}
          </span>
        </div>
        <div className="double-tile__second-value-block">
          <span className="double-tile__value-number">
            {formatSecondValueAsCurrency
              ? formatCurrency(cardSecondValue)
              : cardSecondValue}
          </span>
          <span className="double-tile__subtitle">
            {translate(cardSecondSubtitle, cardDefaultSecondSubtitle)}
          </span>
        </div>
      </div>
      {cardInfo && (
        <div className="double-tile__tooltip">
          <CustomTooltip
            content={translate(cardInfo, cardInfoDefaultValue)}
          >
            <RiInformationLine size={20} />
          </CustomTooltip>
        </div>
      )}
    </div>
  );
};

export default DoubleTile;

DoubleTile.defaultProps = {
  currencyCode: 'USD',
  formatSecondValueAsCurrency: false,
};

DoubleTile.propTypes = {
  cardTitle: PropTypes.string.isRequired,
  cardTitleDefaultValue: PropTypes.string.isRequired,
  cardFirstValue: PropTypes.number.isRequired,
  cardSecondValue: PropTypes.number.isRequired,
  cardFirstSubtitle: PropTypes.string.isRequired,
  cardDefaultFirstSubtitle: PropTypes.string.isRequired,
  cardSecondSubtitle: PropTypes.string.isRequired,
  cardDefaultSecondSubtitle: PropTypes.string.isRequired,
  cardInfo: PropTypes.string.isRequired,
  cardInfoDefaultValue: PropTypes.string.isRequired,
  currencyCode: PropTypes.string,
  formatSecondValueAsCurrency: PropTypes.bool,
};
