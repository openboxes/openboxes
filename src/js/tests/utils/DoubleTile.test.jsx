import React from 'react';

import { render, screen, within } from '@testing-library/react';

import DoubleTile from 'utils/DoubleTile';

import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';

const mockTranslate = jest.fn((key, defaultValue) => defaultValue);

jest.mock('hooks/useTranslate', () => () => mockTranslate);

jest.mock('wrappers/CustomTooltip', () => ({ children }) => <div>{children}</div>);

describe('DoubleTile component', () => {
  const defaultProps = {
    cardTitle: 'react.doubleTile.title',
    cardTitleDefaultValue: 'Card Title',
    cardFirstValue: 75,
    cardSecondValue: 1500,
    cardFirstSubtitle: 'react.doubleTile.firstSubtitle',
    cardDefaultFirstSubtitle: 'First Subtitle',
    cardSecondSubtitle: 'react.doubleTile.secondSubtitle',
    cardDefaultSecondSubtitle: 'Second Subtitle',
    cardInfo: 'react.doubleTile.info',
    cardInfoDefaultValue: 'Info Text',
    currencyCode: 'USD',
    showFirstValuePercentSign: false,
    formatSecondValueAsCurrency: false,
  };

  it('should match snapshot', () => {
    const doubleTile = renderer.create(<DoubleTile {...defaultProps} />).toJSON();
    expect(doubleTile).toMatchSnapshot();
  });

  it('renders title, first value, second value, and subtitles correctly', () => {
    render(<DoubleTile {...defaultProps} />);
    const { getByText } = within(screen.getByTestId('double-tile'));
    expect(getByText('Card Title')).toBeInTheDocument();
    expect(getByText('75')).toBeInTheDocument();
    expect(getByText('First Subtitle')).toBeInTheDocument();
    expect(getByText('1500')).toBeInTheDocument();
    expect(getByText('Second Subtitle')).toBeInTheDocument();
  });

  it('renders first value with percent sign when showFirstValuePercentSign is true', () => {
    render(<DoubleTile {...defaultProps} showFirstValuePercentSign />);
    const { getByText } = within(screen.getByTestId('double-tile'));
    expect(getByText('75%')).toBeInTheDocument();
  });

  it('formats second value as currency when formatSecondValueAsCurrency is true', () => {
    render(<DoubleTile {...defaultProps} formatSecondValueAsCurrency />);
    const { getByText } = within(screen.getByTestId('double-tile'));
    expect(getByText('1,500.00 USD')).toBeInTheDocument();
  });

  it('formats second value as million when value is >= 1,000,000', () => {
    render(<DoubleTile {...defaultProps} cardSecondValue={1500000} formatSecondValueAsCurrency />);
    const { getByText } = within(screen.getByTestId('double-tile'));
    expect(getByText('1.500 million USD')).toBeInTheDocument();
  });

  it('uses translate function for title, subtitles, and tooltip', () => {
    render(<DoubleTile {...defaultProps} />);
    expect(mockTranslate).toHaveBeenCalledWith('react.doubleTile.title', 'Card Title');
    expect(mockTranslate).toHaveBeenCalledWith('react.doubleTile.firstSubtitle', 'First Subtitle');
    expect(mockTranslate).toHaveBeenCalledWith('react.doubleTile.secondSubtitle', 'Second Subtitle');
    expect(mockTranslate).toHaveBeenCalledWith('react.doubleTile.info', 'Info Text');
  });

  it('renders correctly when optional props are undefined', () => {
    const propsWithOptionalDefaults = {
      ...defaultProps,
      currencyCode: undefined,
      showFirstValuePercentSign: undefined,
      formatSecondValueAsCurrency: undefined,
    };
    render(<DoubleTile {...propsWithOptionalDefaults} />);
    const { getByText, queryByText } = within(screen.getByTestId('double-tile'));
    expect(getByText('75')).toBeInTheDocument();
    expect(getByText('1500')).toBeInTheDocument();
    expect(queryByText('1,500.00 USD')).not.toBeInTheDocument();
  });
});
