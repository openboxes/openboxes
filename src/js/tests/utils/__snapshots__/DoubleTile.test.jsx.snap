// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoubleTile component should match snapshot for sample props 1`] = `
<div
  className="double-tile"
  data-testid="double-tile"
>
  <span
    className="double-tile__title"
  >
    Card Title
  </span>
  <div
    className="d-flex"
  >
    <div
      className="double-tile__first-value-block"
    >
      <span
        className="double-tile__value-number"
      >
        75
      </span>
      <span
        className="double-tile__subtitle"
      >
        First Subtitle
      </span>
    </div>
    <div
      className="double-tile__second-value-block"
    >
      <span
        className="double-tile__value-number"
      >
        1500
      </span>
      <span
        className="double-tile__subtitle"
      >
        Second Subtitle
      </span>
    </div>
  </div>
  <div
    className="double-tile__tooltip"
  >
    <div>
      <svg
        fill="currentColor"
        height={20}
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width={20}
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm0-2a8 8 0 1 0 0-16 8 8 0 0 0 0 16zM11 7h2v2h-2V7zm0 4h2v6h-2v-6z"
          />
        </g>
      </svg>
    </div>
  </div>
</div>
`;
